File: src/main/java/com/nikitinho/riskmanager/domain/Risk.java
package com.nikitinho.riskmanager.domain;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonView;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.hibernate.annotations.Type;

import javax.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table
@EqualsAndHashCode(of = {"id"})
@Data
public class Risk {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @JsonView(Views.Id.class)
    private Long id;

    @JsonView(Views.IdName.class)
    @Column(columnDefinition = "TEXT")
    private String text;

    @Column(updatable = false)
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm")
    @JsonView(Views.FullRisk.class)
    private LocalDateTime creationDate;

    @JsonView(Views.FullRisk.class)
    @Column(columnDefinition = "TEXT")
    private String description;

    @JsonView(Views.FullRisk.class)
    @Column(columnDefinition = "TEXT")
    private String causes;

    @JsonView(Views.FullRisk.class)
    @Column(columnDefinition = "TEXT")
    private String consequences;

    @JsonView(Views.FullRisk.class)
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "risk_user",
            joinColumns = {@JoinColumn(name = "risk_id")},
            inverseJoinColumns = {@JoinColumn(name = "user_id")})
    private List<User> responsible;

    @JsonView(Views.FullRisk.class)
    private RiskCategoryType category;

    @JsonView(Views.FullRisk.class)
    private RiskStatusType status;

    @JsonView(Views.FullRisk.class)
    private RiskStageType stage;

    @ManyToOne
    @JoinColumn(name = "project_id")
    @JsonIgnoreProperties("risks")
    @JsonView(Views.FullRisk.class)
    private Project project;

    @JsonView(Views.FullRisk.class)
    private Double probability;

    @JsonView(Views.FullRisk.class)
    private Double impact;

    @JsonView(Views.FullRisk.class)
    private Double riskRate;

    @JsonView(Views.FullRisk.class)
    private RiskLevelType riskLevel;

    @JsonView(Views.FullRisk.class)
    private RiskActionStrategyType strategy;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm")
    @JsonView(Views.FullRisk.class)
    private LocalDateTime actionStartDate;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm")
    @JsonView(Views.FullRisk.class)
    private LocalDateTime actionEndDate;

    @JsonView(Views.FullRisk.class)
    @Column(columnDefinition = "TEXT")
    private String strategyInfo;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm")
    @JsonView(Views.FullRisk.class)
    private LocalDateTime actualActionEndDate;

    @OneToMany(orphanRemoval = true, cascade = {CascadeType.ALL}, mappedBy = "risk")
    @JsonIgnoreProperties("risk")
    @JsonView(Views.FullRisk.class)
    private List<RiskComment> comments;

    public void setRiskLevel(String riskLevel) {
        this.riskLevel = RiskLevelType.fromString(riskLevel);
    }

    public void setRiskLevel(RiskLevelType riskLevel) {
        this.riskLevel = riskLevel;
    }

    public void setStage(String stage) {
        this.stage = RiskStageType.fromString(stage);
    }

    public void setStage(RiskStageType stage) {
        this.stage = stage;
    }

    public void setActionStartDate(String date) {
        if (date == null || date.isEmpty()) {
            this.actionStartDate = null;
        } else {
            LocalDate localDate = LocalDate.parse(date, DateTimeFormatter.ISO_LOCAL_DATE);
            this.actionStartDate = localDate.atTime(23, 59, 59);
        }
    }

    public void setActionStartDate(LocalDateTime date) {
        this.actionStartDate = date;
    }

    public void setActionEndDate(String date) {
        if (date == null || date.isEmpty()) {
            this.actionEndDate = null;
        } else {
            LocalDate localDate = LocalDate.parse(date, DateTimeFormatter.ISO_LOCAL_DATE);
            this.actionEndDate = localDate.atTime(23, 59, 59);
        }
    }

    public void setActionEndDate(LocalDateTime date) {
        this.actionEndDate = date;
    }

    public void setActualActionEndDate(String date) {
        if (date == null || date.isEmpty()) {
            this.actualActionEndDate = null;
        } else {
            LocalDate localDate = LocalDate.parse(date, DateTimeFormatter.ISO_LOCAL_DATE);
            this.actualActionEndDate = localDate.atTime(23, 59, 59);
        }
    }

    public void setActualActionEndDate(LocalDateTime date) {
        this.actualActionEndDate = date;
    }

    public void setComments(List<RiskComment> comments) {
        if (this.comments == null) {
            this.comments = new ArrayList<>();
            this.comments.addAll(comments);
        } else {
            this.comments.clear();
            this.comments.addAll(comments);
        }
    }
}

File: src/main/resources/static/js/domain/Risk.js
// Risk
// All fields coincide with .java class!
// All enums coincide with .java enums!
export default class Risk {

    constructor (risk) {
        if (risk) {
            this.id = risk.id
            this.text = risk.text
            this.creationDate = risk.creationDate
            this.description = risk.description
            this.category = risk.category
            this.causes = risk.causes
            this.consequences = risk.consequences
            this.responsible = risk.responsible
            this.status = risk.status
            this.stage = Risk.convertStage(risk.stage)
            this.probability = risk.probability
            this.impact = risk.impact
            this.riskRate = risk.riskRate
            this.riskLevel = Risk.convertRiskLevel(risk.riskLevel)
            this.strategy = risk.strategy
            this.actionStartDate = risk.actionStartDate ? risk.actionStartDate.substr(0, 10) : null
            this.actionEndDate = risk.actionEndDate ? risk.actionEndDate.substr(0, 10) : null
            this.strategyInfo = risk.strategyInfo
            this.actualActionEndDate = risk.actualActionEndDate ? risk.actualActionEndDate.substr(0, 10) : null
            this.comments = risk.comments || []
        } else {
            this.text = ''
            this.creationDate = ''
            this.description = ''
            this.category = null
            this.causes = []
            this.consequences = ''
            this.responsible = []
            this.status = null
            this.stage = null
            this.probability = ''
            this.impact = ''
            this.riskRate = ''
            this.riskLevel = null
            this.strategy = null
            this.actionStartDate = null
            this.actionEndDate = null
            this.strategyInfo = ''
            this.actualActionEndDate = null
            this.comments = []
        }
    }

    static convertRiskLevel(level) {
        switch (level) {
            case 'LOW':
                return 'Низкий'
            case 'MEDIUM':
                return 'Средний'
            case 'HIGH':
                return 'Высокий'
            default:
                return null
        }
    }

    static convertStage(stage) {
        switch (stage) {
            case 'IDENTIFICATION':
                return 'Идентификация'
            case 'EVALUATION':
                return 'Оценка'
            case 'PLANNING':
                return 'Планирование'
            case 'MONITORING':
                return 'Мониторинг'
            default:
                return null
        }
    }

    static convertStageToStep(stage) {
        switch (stage) {
            case 'IDENTIFICATION':
                return 1
            case 'EVALUATION':
                return 2
            case 'PLANNING':
                return 3
            case 'MONITORING':
                return 4
            default:
                return 0
        }
    }

    static convertRiskRateToColor(level) {
        if (level === 0.0) {
            return 'grey'
        } else if (level > 0.0 && level < 0.34) {
            return 'green'
        } else if (level >= 0.34 && level < 0.68) {
            return 'yellow'
        } else if (level >= 0.68 && level <= 1.0) {
            return 'red'
        }
    }

    static convertRiskRateToLevel(rate) {
        if (rate >= 0.0 && rate < 0.34) {
            return 'Низкий'
        } else if (rate >= 0.34 && rate < 0.68) {
            return 'Средний'
        } else if (rate >= 0.68 && rate <= 1.0) {
            return 'Высокий'
        }
    }
}